Neo4j-Browser becomes unresponsive with large query results

huboneo opened the pull request:
This PR introduces a new configuration option in browser for truncating max field length before they are passed to rendering layers. The reason for this is that certain queries can produce a single result with a lot of values in a given field, which we had yet to account for.

### Screenshot
<img width="1192" alt="Screenshot 2019-12-12 at 13 01 29" src="https://user-images.githubusercontent.com/52443771/70712645-85bb7e80-1ce4-11ea-85d7-e9f5dc35290b.png">
<img width="1635" alt="Screenshot 2020-01-07 at 12 30 25" src="https://user-images.githubusercontent.com/52443771/71892843-50bb1400-314a-11ea-9b91-2d94054027a5.png">

huboneo commented on src/shared/services/bolt/bolt.js:
```diff
@@ -288,6 +288,8 @@ export default {
   ) => {
     const intChecker = neo4j.isInt
     const intConverter = val => val.toString()
+    return { nodes: [], relationships: [] }
```
Remove this, WIP

huboneo committed:
Best effort performance fix for large result fields
- Introduced new setting "maxFieldItems"
- Now truncating data passed to CypherFrames whenever a field that is too large is detected
- Refactored boltMapping code to increase performance
- Side effect is that we now dedupe nodes, not sure if OK

huboneo committed:
bugfix unit test

huboneo committed:
fix performance of flattenArray

huboneo committed:
update snapshots, add warning when fields are truncated

oskarhane reviewed (commented):
Some really good stuff in here.

Some comments: 

- I get somewhat inconsistent results. If you have the movies database and set the "Max field items" to `1` and execute `MATCH p=()-[r:ACTED_IN]->() RETURN p LIMIT 25` I see two nodes in the viz. 
But switching to table view I see 25 node objects.

- The limitations we already have with "Max Rows" in table view + "Initial node display" in viz, maybe those should be removed and replaced by this one (since this operates at a lower level)?

- I don't see a truncation message in the viz's statusbar.  

- Maybe not part of this PR, but just to point it out, property size is also a performance killer (try returning a node with a property of 1MB).

huboneo committed:
changes from review

huboneo commented:
@oskarhane thank you for your feedback.
- Indeed there was a bug truncation.
- Added a message to Vis for truncated messages
- Updated setting display name
- Unsure about removing max row truncation as that is on a different axis from fields. Lets leave it as is for now?

huboneo commented on src/browser/modules/Stream/CypherFrame/CodeView.jsx:
```diff
@@ -54,16 +57,33 @@ class ExpandableContent extends Component {
+      request.result.records,
+      fieldLimiterFactory(maxFieldItems),
+      2
+    )
```
Unfortunately not, even `JSON.stringify()` gets overwhelmed with large result fields

oskarhane commented on src/browser/modules/Stream/CypherFrame/TableView.jsx:
```diff
@@ -190,14 +201,30 @@ export class TableStatusbar extends Component {
+        <Ellipsis>
+          {this.state.hasTruncatedFields && (
+            <StyledTruncatedMessage>
+              <Icon name='warning sign' /> Result fields have been
```
"Result fields" => "Records fields"

oskarhane commented on src/browser/modules/Sidebar/Settings.jsx:
```diff
@@ -159,6 +159,13 @@ const visualSettings = [
       },
+      {
+        maxFieldItems: {
+          displayName: 'Property item limit',
```
"Record field limit" describes this better imo. wdyt?

oskarhane commented on src/browser/modules/Sidebar/Settings.jsx:
```diff
@@ -159,6 +159,13 @@ const visualSettings = [
+        maxFieldItems: {
+          displayName: 'Property item limit',
+          tooltip:
+            'Max amount of items shown for list properties. When reached, lists get truncated.'
```
"Max number of fields per record being handled" os more clear. wdyt?

oskarhane commented on src/browser/modules/D3Visualization/components/Inspector.jsx:
```diff
@@ -149,6 +152,12 @@ export class InspectorComponent extends Component {
               <StyledInspectorFooterRowListValue className="value">
+                {this.props.hasTruncatedFields && (
+                  <StyledTruncatedMessage>
+                    <Icon name="warning sign" /> Result fields have been
```
"Result fields" => "Records fields"

oskarhane reviewed (commented):
Thanks for the update on this.

## Result bug
I still get mismatches between the different view though.
To reproduce:

1. Load the common movie dataset
1. Set the limit to 1
1. Run `MATCH p=(n:Movie)-[]-(:Person) RETURN p LIMIT 25` and compare the views. The table view only sees one item per row while viz and text show the whole path.
See:
<img width="1038" alt="Screenshot 2020-01-13 09 27 13" src="https://user-images.githubusercontent.com/570998/72241436-243e4680-35e7-11ea-9d2d-f5693ce0480d.png">

## Phrasing
I've re-tought the phrasing as well and I think it's ok to use the word "fields" as this is the first time we describe the records width to users. So sorry about the back-and-forth on that.
I've suggest what wording to use via comments.

huboneo committed:
only truncate top level cell values in tables

huboneo committed:
wording

huboneo commented:
@oskarhane thank you again for your feedback. I have updated the PR. Tests are still failing I believe

huboneo commented:
<img width="1060" alt="Screenshot 2020-01-13 at 10 51 47" src="https://user-images.githubusercontent.com/52443771/72246313-b861db00-35f2-11ea-9ef2-4080cbc3604b.png">

oskarhane reviewed (commented):
Thanks for the update!

I still see different results in table vs text view. They should be the same. The viz does some magic with internal relationships, but not the different table views.

1. Load movie dataset
1. Set limit to 1
1. Run `MATCH p=()-[r:FOLLOWS]->() RETURN p LIMIT 25` and compare table and text views.

<img width="1117" alt="Screenshot 2020-01-13 12 23 11" src="https://user-images.githubusercontent.com/570998/72252556-c5d19200-35ff-11ea-971d-a9c77fc8d6ff.png">


👍 on the wording changes.

oskarhane commented:
Stable version of react-table is now out @huboneo. When yo have time, could you update this upstream and update this PR?

huboneo commented:
closing in favour of https://github.com/neo4j/neo4j-browser/pull/1090
